### ISSUE TITLE
Beam.density segfaults when called without Beam.attenuator

### ISSUE THREAD

**Issue by Mateasek** [cherab/core#286]
Segmentation fault is caused when the Beam.density method is called without setting the Beam.attenuator. This can happen when there are no emission models assigned, because all the checks happen when models are added. There are two options to fix this I think:

1) Add a check into the density method if the attenuator is None and throw an error. The negative side of this solution is that the check is not needed when called by the emission models, which is the majority of calls.

2) Set up a "ZeroAttenuator" as a default value which returns zero density. The plus of this is that the density method stays untouched and errors could still be raised when an emission model is being set before the attenuator.

I personally prefer the 2nd option.

**Comment by vsnever** [comment]
Hi @Mateasek,

I think that setting up a `SingleRayAttenuator` as a default value is better. It looks safe despite the fact that the attenuator setter requires setting the plasma and atomic data before the attenuator. All public methods of `SingleRayAttenuator` populate the stopping cache first. If the plasma or atomic data is missing, the error will be raised already at this stage. Both the plasma and atomic data setters update the attenuator, so it's seem ok to set them after the attenuator is already set up. 

In a situation where the user has forgotten to set up the attenuator, a `ZeroAttenuator` will return zero emission or density, the cause of which may be difficult to determine. In the same situation, a `SingleRayAttenuator` will return correct values.

However, I haven't tested this and may be wrong, so please check.

**Comment by Mateasek** [comment]
Hi @vsnever,

I also thought about this solution, but the error output could be misleading if the SingleRayAttenuator is set, couldn't it? Also I wasn't really comfortable with assigning a particular model as a default one. What about if the density method of the ZeroAttenuator raised an error? In that case it would be obvious where is the problem.

**Comment by vsnever** [comment]
The `SingleRayAttenuator` is the only attenuator model currently implemented in Cherab, therefore it seems natural to use it as a default one. At the same time, if for example the integration step of `SingleRayAttenuator` is not optimal for a given beam and plasma, the output can be inaccurate, which can be really misleading. 

> What about if the density method of the ZeroAttenuator raised an error? In that case it would be obvious where the problem is.

Yes, I think it's better than returning zero density.

**Comment by jacklovell** [comment]
Why not just make the default attenuator a `BeanAttenuator` instance? That does what you require: it raises `NotImplementedError` when the density function is called. All attenuators must be instances of this base class anyway, so rather than implement a new "do nothing" subclass you may as well use the "do nothing" base class.

**Comment by Mateasek** [comment]
This is also one option I had in mind, the only disadvantage I see is it would make it less straightforward to detect if it was initialised by user on not, because if you make it `BeamAttenuator` then you can't simply use `isinstance` method as usual, can you? Some of the models might want to implement this check.

**Comment by jacklovell** [comment]
All other implementations (including the proposed `ZeroAttenuator`) must be instances of `BeamAttentuator` anyway. So `isinstance(model, BeamAttenuator)` will always return true. If other models want finer control over how they treat different attenuator models they should be using `isinstance(model, <some subclass of BeamAttenuator>`, which will return `False` if the default `BeamAttenuator` instance is used anyway.

**Comment by Mateasek** [comment]
True

**Comment by Mateasek** [comment]
If you want to prevent the emission models to be assigned without the user defining the attenuator [here](https://github.com/cherab/core/blob/19b5237e97231365063eef2f181ecdfbca4f8bce/cherab/core/beam/node.pyx#L423) then you have no simple way how to identify, do you? On the other hand if you ommit this you are right that this is covered by the `BeamAttenuator` raising an error when used.

**Comment by vsnever** [comment]
In any case, if attenuator is set by default prior to plasma and atomic_data, the `_configure_attenuator()` must be changed to something like this:
```python
    def _configure_attenuator(self):

        # there must be an attenuator present to configure
        if not self._attenuator:
            return

        if self._plasma is not None:
            self._attenuator.plasma = self._plasma

        if self._atomic_data is not None:
            self._attenuator.atomic_data = self._atomic_data

        self._attenuator.beam = self
```
Otherwise, it will throw an error when the user tries to set up plasma or atomic data.

**Comment by jacklovell** [comment]
Arguably the check you reference wasn't strong enough, as it's still possible to crash the program bypassing that check. I think it would be better to consider what specific problem not having a user-defined attenuator causes when using the models? If it's that the density function will raise, then check whether `self.attenuator.density` raises.

There are a few instances of `if not self._attenuator` in the node class. Mostly it looks like they're there to avoid trying to call `BeamAttenuator` methods on a `None` object and segfaulting. That won't be necessary if `self._attenuator` is always guaranteed to be a `BeamAttenuator` instance.

**Comment by jacklovell** [comment]
> In any case, if attenuator is set by default prior to plasma and atomic_data, the `_configure_attenuator()` must be changed to something like this:
> 
> ```python
>     def _configure_attenuator(self):
> 
>         # there must be an attenuator present to configure
>         if not self._attenuator:
>             return
> 
>         if self._plasma is not None:
>             self._attenuator.plasma = self._plasma
> 
>         if self._atomic_data is not None:
>             self._attenuator.atomic_data = self._atomic_data
> 
>         self._attenuator.beam = self
> ```
> 
> Otherwise, it will throw an error when the user tries to set up plasma or atomic data.

None of this will throw an error if `self._attenuator` is an instance of `BeamAttenuator`. The base class has all of these required properties already defined and knows how to handle them.

**Comment by jacklovell** [comment]
Oh, I see what you mean: `_configure_attenuator` may be called before `self._plasma` and `self._atomic_data` are both set.

**Comment by vsnever** [comment]
> Oh, I see what you mean: _configure_attenuator may be called before self._plasma and self._atomic_data are both set.

Yes. It is called in `plasma` and `atomic_data` setters, assuming that `_attenuator` is not set up yet.

**Comment by jacklovell** [comment]
In which case the check for `None` is sensible. It chould happen in `configure_attenuator` as you've proposed, or in the setters for plasma and atomic_data. At the moment, passing e.g. `self._attenuator.plasma = None` will set `self._plasma` to None but then error out trying to access `self._plasma.notifier.add`. That's probably a bug too. Same with the atomic_data setter.

**Comment by vsnever** [comment]
> At the moment, passing e.g. self._attenuator.plasma = None will set self._plasma to None but then error out trying to access self._plasma.notifier.add. That's probably a bug too. Same with the atomic_data setter.

Oh, yes it is. The attenuator setters for `atomic_data` and `plasma` must be updated to accept only `AtomicData` and `Plasma` instances, just like those in `Beam`: 
`def atomic_data(self, AtomicData value not None):`, 
`def plasma(self, Plasma value not None):`.

**Comment by vsnever** [comment]
`_generate_geometry()` will crash if attenuator does not have `clamp_sigma` attribute, https://github.com/cherab/core/blob/19b5237e97231365063eef2f181ecdfbca4f8bce/cherab/core/beam/node.pyx#L488 
And the base `BeamAttenuator` class does not have this attribute, so this is another bug.

**Comment by Mateasek** [comment]
So after all wouldn't it be simpler to add the `ZeroBeamAttenuator` and change the `if self._attenuator is None` and similar to `if isinstance(self._attenuator, ZeroBeamAttenuator)`? To me it now seems like the simplest and less intrusive solution. I would anyway like to add another version of the beam attenuator, so when this happens some of the problems, like the `num_sigma = self._attenuator.clamp_sigma` can be removed.

**Comment by Mateasek** [comment]
Also, in future the beam geometry might have to be driven by the attenuator to get a better coverage of the beam. This would mean that some parts of the Beam node will have to be changed anyway.

**Comment by CnlPepper** [comment]
This seems like a lot of work to avoid a `if blah is None` check. Basic checks for None have literally no impact on performance if the result is the same in the majority of cases (which it will be) - the CPU branch predictor will rapidly learn the most probable path. Just check for None and raise an exception. There are already loads of None checks for caching already that have no impact.

**Comment by Mateasek** [comment]
The slowdown is small, that is true. Also the emission models can avoid the check by getting the beam density from the attenuator directly instead of calling the Beam.density, if they need. I just thought it would be nice to avoid adding another check which was redundant in majority of calls. I will add the check and push the fix.

**Comment by vsnever** [comment]
In addition to @Mateasek's PR #287, I'm going to fix the setter bugs we found in `BeamModel` and `BeamAttenuator` before we forgot about them. Do I also need to move `clamp_sigma` and `clamp_to_zero` properties from the `SingleRayAttenuator` to the base `BeamAttenuator`? These properties look quite general. Doing so will prevent `Beam._generate_geometry()` from crashing when the attenuator is not an instance of `SingleRayAttenuator`.

**Comment by Mateasek** [comment]
@vsnever the state we have now has been there for years and as I tried to express it in the post above I would like to fix this as a part of a larger change. I think the best would be to move the geometry methods to the attenuator. I will make a separate issue about it.

**Comment by vsnever** [comment]
Thanks. Then I'll fix only the setters.

**Comment by Mateasek** [comment]
fixed with PR #287
